### PR TITLE
ISPN-13754 BlockingManager should not blocking invoking thread if sta…

### PR DESCRIPTION
**Pull request by wburns**
…ge is not complete

https://issues.redhat.com/browse/ISPN-13754
https://issues.redhat.com/browse/ISPN-13758

### CODE REVIEW

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
any chance of adding `thenComposeBlocking` to this PR? :)

**Comment by wburns** [comment]
> any chance of adding `thenComposeBlocking` to this PR? :)

Sure, let me make another JIRA and I can add this.

**Comment by wburns** [comment]
@pruivo added!

File: core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java [hunk 1]
```diff
@@ -207,10 +207,22 @@ public void execute(Runnable command) {
...
    }
 
+   @Override
+   public <I, O> CompletionStage<O> thenComposeBlocking(CompletionStage<? extends I> stage,
+         Function<? super I, ? extends CompletionStage<O>> function, Object traceId) {
+      if (CompletionStages.isCompletedSuccessfully(stage) && isCurrentThreadBlocking()) {
+         if (log.isTraceEnabled()) {
+            log.tracef("Invoked thenComposeBlocking on a blocking thread, joining %s in same blocking thread", traceId);
+         }
+         return stage.thenCompose(function);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:217]: you probably could do a `join()` and invoke `function.apply()` directly.

File: core/src/main/java/org/infinispan/util/concurrent/BlockingManager.java [hunk 2]
```diff
@@ -95,8 +95,9 @@
     * (if the current thread is non-blocking) or in the current thread (if the current thread is blocking).
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/BlockingManager.java:95]: missing `and the stage is completed` from all the modified methods :)

File: core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:217]: Sure, should be slightly less allocations for only marginally more code.

**Comment by wburns** [comment]
Updated

File: core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java [hunk 3]
```diff
@@ -176,7 +176,7 @@ public void execute(Runnable command) {
    @Override
    public <I, O> CompletionStage<O> thenApplyBlocking(CompletionStage<? extends I> stage,
          Function<? super I, ? extends O> function, Object traceId) {
-      if (isCurrentThreadBlocking()) {
+      if (CompletionStages.isCompletedSuccessfully(stage) && isCurrentThreadBlocking()) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:179]: BTW, wouldn't make more sense to check if it is completed, successfully or not, and handle exceptions?

File: core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java [hunk 4]
```diff
@@ -207,10 +207,23 @@ public void execute(Runnable command) {
...
 
+   @Override
+   public <I, O> CompletionStage<O> thenComposeBlocking(CompletionStage<? extends I> stage,
+         Function<? super I, ? extends CompletionStage<O>> function, Object traceId) {
+      if (CompletionStages.isCompletedSuccessfully(stage) && isCurrentThreadBlocking()) {
+         if (log.isTraceEnabled()) {
+            log.tracef("Invoked thenComposeBlocking on a blocking thread, joining %s in same blocking thread", traceId);
+         }
+         I value = CompletionStages.join(stage);
+         return function.apply(value);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:218]: `try { ... }  catch { ... }` missing IMO

Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:218]: Ah yes it is now that I am not using thenCompose anymore :(

File: core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:179]: I had debated about that, but I personally don't feel it is really worth it to worry about exceptions for the invoking blocking thread path. If anything I would say we can change that in a separate PR if we really wanted.

**Comment by pruivo** [comment]
merged! thanks @wburns !

**Pull request merged**